pinot-core/src/main/java/com/linkedin/pinot/core/plan/maker/PlanNodeFactory.java

```diff
@@ -0,0 +1,161 @@
+/**
+ * Copyright (C) 2014-2016 LinkedIn Corp. (pinot-core@linkedin.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.linkedin.pinot.core.plan.maker;
+
+import java.util.List;
+
+import com.linkedin.pinot.common.request.AggregationInfo;
+import com.linkedin.pinot.common.request.BrokerRequest;
+import com.linkedin.pinot.common.segment.SegmentMetadata;
+import com.linkedin.pinot.core.indexsegment.IndexSegment;
+import com.linkedin.pinot.core.plan.AggregationGroupByPlanNode;
+import com.linkedin.pinot.core.plan.AggregationPlanNode;
+import com.linkedin.pinot.core.plan.DictionaryBasedAggregationPlanNode;
+import com.linkedin.pinot.core.plan.MetadataBasedAggregationPlanNode;
+import com.linkedin.pinot.core.plan.PlanNode;
+import com.linkedin.pinot.core.plan.SelectionPlanNode;
+import com.linkedin.pinot.core.query.aggregation.function.AggregationFunctionFactory;
+import com.linkedin.pinot.core.segment.index.ColumnMetadata;
+import com.linkedin.pinot.core.segment.index.SegmentMetadataImpl;
+
+public class PlanNodeFactory {
```

Comment:
whats the idea behind introducing a plannodeFactory.

Reply:
Mayank made a suggestion to introduce a factory to return the right plan node

Reply:
Yes, my suggestion was to factor out the nested inlined 'ifs', and perhaps use factory for the same. This implementation does not adhere to factory pattern per say. The 'isFitFor' functions don't necessarily belong here, they could be either in the original class, or some utility class. Factory can just take the type of object to return.